src/qumin/representations/patterns.py

```diff
@@ -871,152 +1050,109 @@ def find_patterns_in_col(column, pat_dict):
             for p in collection[alt]:
-                p.score = _score(p, (c1, c2), forms, index)
+                p.score = _score(p, df)
                 sorted_collection.append(p)
 
         # Sort by score and choose best patterns
         sorted_collection = sorted(sorted_collection, key=lambda x: x.score, reverse=True)
-        best = defaultdict(set)
-
-        nullset = {''}
-
-        for l in index:
-            row = col.at[l, c1], col.at[l, c2]
 
-            if row != (nullset, nullset):
-                if row[0] == row[1]:  # If the lists are identical, do not compute product
-                    pairs = [(x, x) for x in row[0]]
-                else:
-                    pairs = product(*row)
-
-                for a, b in pairs:
-                    best[l].add(attribute_best_pattern(sorted_collection, l, a, b))
-            else:
-                best[l].add(None)
+        def _best_pattern(row):
+            lex, a, b = row.lexeme, row.form_x, row.form_y
+            if a != '' and b != '':
+                return _attribute_best_pattern(sorted_collection, lex, a, b)
+            return None
 
+        df['pattern'] = df.apply(_best_pattern, axis=1)
         if optim_mem:
-            result = [repr(PatternCollection(list(best[l]))) for l in index]
+            df.pattern = df.pattern.apply(repr)
         else:
-            result = [PatternCollection(list(best[l])) for l in index]
-            pat_dict[(c1, c2)] = list(set.union(*[set(coll.collection) for coll in result]))
-
-        return result
-
-    tqdm.pandas(leave=False, disable=disable_tqdm)
-    patterns_df = patterns_df.progress_apply(find_patterns_in_col, axis=0, args=(pat_dict,))
-
-    return patterns_df, pat_dict
-
-
-def find_applicable(paradigms, pat_dict, disable_tqdm=False, **kwargs):
-    """Find all applicable rules for each form.
-
-    We name sets of applicable rules *classes*. *Classes* are oriented:
-    we produce two separate columns (a, b) and (b, a)
-    for each pair of columns (a, b) in the paradigm.
-
-    Arguments:
-        paradigms (:class:`pandas:pandas.DataFrame`):
-            paradigms (columns are cells, index are lemmas).
-        pat_dict  (dict): a dict mapping a column name to a list of patterns.
-        disable_tqdm (bool): if true, do not show progressbar
-
-    Returns:
-        :class:`pandas:pandas.DataFrame`:
-            associating a lemma (index)
-            and an ordered pair of paradigm cells (columns)
-            to a tuple representing a class of applicable patterns.
-    """
-
-    def _iter_applicable_patterns(form, local_patterns, cell):
-        known_regexes = set()
-        if type(form) is tuple:  # if overabundant
-            form = form[0]  # from tuple to Form
-        for pattern in local_patterns:
-            regex = pattern._regex[cell]
-            if regex in known_regexes:
-                yield pattern
+            self.pat_dict[(c1, c2)] = df['pattern'].unique()
+        return (pair, df)
```

Comment:
Note that I had to change the return type here so that we know where to fit the results, which come unordered